src/lib/string_pool.rs

```diff
@@ -0,0 +1,506 @@
+    }
+
+    /// Resets the current bump vec to the beginning
+    pub(crate) fn clear_current(&self) {
+        self.set_len(0)
+    }
+
+    /// Moves the start pointer to the current length so that a new bump vec region begins
+    /// REVIEW: Maybe shouldn't be necessary or just force a new vec creation..?
+    pub(crate) fn finish_current(&self) {
+        self.consume_current_vec();
+    }
+
+    // REVIEW: Not sure if necessary
+    pub(crate) fn finish_string(&self) -> &mut [XML_Char] {
+        self.consume_current_vec()
+    }
+
+    /// Obtains the length of the current BumpVec.
+    pub(crate) fn len(&self) -> usize {
+        self.inner().rent(|vec| vec.borrow().len())
+    }
+
+    /// Call callback with an immutable buffer of the current BumpVec. This must
+    /// be a callback to ensure that we don't (safely) borrow the slice for
+    /// longer than it stays vaild.
+    pub(crate) fn current_slice<F, R>(&self, mut callback: F) -> R
+        where F: FnMut(&[XML_Char]) -> R
+    {
+        self.inner().rent(|v| callback(v.borrow().as_slice()))
+    }
+
+    /// Unsafe temporary version of `current_slice()`. This needs to be removed
+    /// when callers are made safe.
+    pub(crate) unsafe fn current_start(&self) -> *const XML_Char {
+        self.inner().rent(|v| v.borrow().as_ptr())
+    }
+
+    /// Appends a char to the current BumpVec.
+    pub(crate) fn appendChar(&self, c: XML_Char) -> bool {
+        if self.is_full() && !self.grow() {
+            false
+        } else {
+            self.inner().rent(|buf| buf.borrow_mut().push(c));
+
+            true
+        }
+    }
+
+    /// Overwrites the last char in the current BumpVec.
+    /// Note that this will panic if empty and that this is not an insert
+    /// operation as it does not shift bytes afterwards.
+    pub(crate) fn prepend_char(&self, c: XML_Char) {
+        self.inner().rent(|buf| {
+            *buf
+                .borrow_mut()
+                .last_mut()
+                .expect("Called prepend_char() when string was empty") = c;
+        })
+    }
+
+    /// Decrements the length, panicing if len is 0 in debug
+    pub(crate) fn backtrack(&self) {
+        self.inner().rent(|vec| vec.borrow_mut().pop().expect("Called backtrack() on empty BumpVec"));
+    }
+
+    /// Gets the last character, panicing if len is 0
+    pub(crate) fn get_last_char(&self) -> XML_Char {
+        self.inner().rent(|buf| *buf.borrow().last().expect("Called get_last_char() when string was empty"))
+    }
+
+    /// Appends an entire C String to the current BumpVec.
+    pub(crate) unsafe fn appendString(&self, mut s: *const XML_Char) -> bool {
+        while *s != 0 {
+            if !self.appendChar(*s) {
+                return false;
+            }
+            s = s.offset(1)
+        }
+        true
+    }
+
+    /// Resets the current Bump and deallocates its contents.
+    pub(crate) fn clear(&mut self) {
```

Comment:
I keep thinking we should re-architect this to take an owned `self` and return a new `Self` with a new lifetime. We should be invalidating all references to the string pool here, or this isn't safe. Could we do this and make the owners an `Option<StringPool>` and make the owner replace it?

Reply:
In fact, this isn't safe as is along with `consume_current_vec()` because it invalidates slice references that should be valid for the lifetime of the string pool, without ending that lifetime.

Reply:
`clear` has exclusive access, which should conflict with any existing immutable references right? For example this code fails to compile:

```Rust
#[test]
fn test_clear() {
    use consts::*;

    let mut pool = StringPool::try_new().unwrap();

    unsafe {
        pool.appendString(S.as_ptr());
    }

    pool.current_slice(|s| assert_eq!(s, [C, D, D, C]));

    let string = pool.finish_string();

    pool.current_slice(|s| assert!(s.is_empty()));

    assert_eq!(string, [C, D, D, C]);

    pool.clear();

    assert_eq!(string, [C, D, D, C]);
}
```

```Rust
error[E0502]: cannot borrow `pool` as mutable because it is also borrowed as immutable
   --> src/lib/string_pool.rs:519:5
    |
513 |     let string = pool.finish_string();
    |                  ---- immutable borrow occurs here
...
519 |     pool.clear();
    |     ^^^^^^^^^^^^ mutable borrow occurs here
520 |
521 |     assert_eq!(string, [C, D, D, C]);
    |     --------------------------------- immutable borrow later used here
    |
```

Reply:
So you can't call `clear` while there are still lingering string references. Obviously, raw pointers throw a wrench into this, but that's a separate issue.

Reply:
Yeah, you're right, I was forgetting that references keep the pool itself borrowed.